File: NetDeck/CardTypeStats.m
//
//  CardTypeStats.m
//  Net Deck
//
//  Created by Gereon Steffens on 17.02.14.
//  Copyright © 2016 Gereon Steffens. All rights reserved.
//

#import "CardTypeStats.h"

@interface CardTypeStats()
@property NSInteger deckSize;
@end

@implementation CardTypeStats

-(CardTypeStats*) initWithDeck:(Deck *)deck
{
    if ((self = [super init]))
    {
        // calculate type distribution
        NSMutableDictionary* types = [NSMutableDictionary dictionary];
        for (CardCounter* cc in deck.cards)
        {
            NSString* type = l10n(cc.card.typeStr);
            
            if (cc.card.type == NRCardTypeProgram) {
                type = cc.card.programType;
            }
            NSNumber* n = [types objectForKey:type];
            int prev = n == nil ? 0 : [n intValue];
            n = @(prev + cc.count);
            [types setObject:n forKey:type];
        }
        
        NSArray* sections = [[types allKeys] sortedArrayUsingComparator:^(NSString* s1, NSString* s2) {
            return [s1 compare:s2];
        }];
        NSMutableArray* values = [NSMutableArray array];
        for (NSString*s in sections)
        {
            [values addObject:[types objectForKey:s]];
        }
        NSAssert(sections.count == values.count, @"");
        self.tableData = [[TableData alloc] initWithSections:sections andValues:values];
        self.deckSize = deck.size;
    }
    return self;
}

-(CPTGraphHostingView*) hostingView
{
    return [self hostingViewForDelegate:self identifier:@"Card Type"];
}

#pragma mark - CPTPlotDataSource methods

-(CPTLayer *)dataLabelForPlot:(CPTPlot *)plot recordIndex:(NSUInteger)index
{
    static CPTMutableTextStyle *labelText = nil;
    
    if (!labelText)
    {
        labelText = [[CPTMutableTextStyle alloc] init];
        labelText.color = [CPTColor blackColor];
    }
    
    NSString* type = [self.tableData.sections objectAtIndex:index];
    NSNumber* cards = [self.tableData.values objectAtIndex:index];
    
    NSString* str = nil;
    if ([cards intValue] > 0)
    {
        float pct = [cards intValue] * 100.0 / _deckSize;
        
        str = [NSString stringWithFormat:@"%@: %d\n%.1f%%", type, [cards intValue], pct];
    }
    
    // 5 - Create and return layer with label text
    return [[CPTTextLayer alloc] initWithText:str style:labelText];
}

@end

File: NetDeck/CardTypeStats.swift
//
//  CardTypetats.swift
//  NetDeck
//
//  Created by Gereon Steffens on 25.12.16.
//  Copyright © 2017 Gereon Steffens. All rights reserved.
//

import CorePlot

class CardTypeStats: Stats {

    private let deckSize: Int
    
    init(deck: Deck) {
        self.deckSize = deck.size
        super.init()
        
        var types = [String: Int]()
        for cc in deck.cards {
            let card = cc.card
            let type = card.type == .program ? card.programType : card.typeStr.localized()
            
            types[type] = (types[type] ?? 0) + cc.count
        }
        
        let sections = types.keys.sorted { $0 < $1 }
        let values = sections.map { [types[$0]!] }
        
        assert(sections.count == values.count)
        
        self.tableData = TableData(sections: sections, values: values)
    }
    
    var hostingView: CPTGraphHostingView {
        return self.hostingView(for: self, identifier: "Card Type")
    }
    
    func dataLabel(for plot: CPTPlot, record index: UInt) -> CPTLayer? {
        let type = self.tableData.sections[Int(index)]
        let cards = self.tableData.values[Int(index)][0]
        
        let pct = Float(cards) * 100.0 / Float(self.deckSize)
        
        let str = String(format: "%@: %d\n%.1f%%", type, cards, pct)
        return CPTTextLayer(text: str)
    }
    
}
